pushsync/simulation_test.go

```diff
@@ -0,0 +1,295 @@
+
+package pushsync
+
+import (
+	"context"
+	"encoding/hex"
+	"flag"
+	"fmt"
+	"io/ioutil"
+	"math/rand"
+	"os"
+	"path/filepath"
+	"sync"
+	"testing"
+	"time"
+
+	"github.com/ethereum/go-ethereum/crypto"
+	"github.com/ethereum/go-ethereum/node"
+	"github.com/ethereum/go-ethereum/p2p"
+	"github.com/ethereum/go-ethereum/p2p/simulations/adapters"
+	"github.com/ethersphere/swarm/chunk"
+	"github.com/ethersphere/swarm/log"
+	"github.com/ethersphere/swarm/network"
+	"github.com/ethersphere/swarm/network/simulation"
+	"github.com/ethersphere/swarm/network/stream"
+	"github.com/ethersphere/swarm/pss"
+	"github.com/ethersphere/swarm/state"
+	"github.com/ethersphere/swarm/storage"
+	"github.com/ethersphere/swarm/storage/localstore"
+)
+
+var (
+	bucketKeyPushSyncer = simulation.BucketKey("pushsyncer")
+	bucketKeyNetStore   = simulation.BucketKey("netstore")
+)
+
+var (
+	nodeCntFlag   = flag.Int("nodes", 4, "number of nodes in simulation")
+	chunkCntFlag  = flag.Int("chunks", 4, "number of chunks per upload in simulation")
+	testCasesFlag = flag.Int("cases", 4, "number of concurrent upload-download cases to test in simulation")
+)
+
+// test syncer using pss
+// the test
+// * creates a simulation with connectivity loaded from a snapshot
+// * for each test case, two nodes are chosen randomly, an uploader and a downloader
+// * uploader uploads a number of chunks
+// * wait until the uploaded chunks are synced
+// * downloader downloads the chunk
+// Testcases are run concurrently
+func TestPushsyncSimulation(t *testing.T) {
+	nodeCnt := *nodeCntFlag
+	chunkCnt := *chunkCntFlag
+	testcases := *testCasesFlag
+
+	err := testSyncerWithPubSub(nodeCnt, chunkCnt, testcases, newServiceFunc)
+	if err != nil {
+		t.Fatal(err)
+	}
+}
+
+func testSyncerWithPubSub(nodeCnt, chunkCnt, testcases int, sf simulation.ServiceFunc) error {
+	sim := simulation.NewInProc(map[string]simulation.ServiceFunc{
+		"pushsync": sf,
+	})
+	defer sim.Close()
+
+	ctx := context.Background()
+	snapCtx, cancel := context.WithTimeout(ctx, 30*time.Second)
+	defer cancel()
+	err := sim.UploadSnapshot(snapCtx, filepath.Join("../network/stream/testing", fmt.Sprintf("snapshot_%d.json", nodeCnt)))
+	if err != nil {
+		return fmt.Errorf("error while loading snapshot: %v", err)
+	}
+
+	start := time.Now()
+	log.Info("Snapshot loaded. Simulation starting", "at", start)
+	result := sim.Run(ctx, func(ctx context.Context, sim *simulation.Simulation) error {
+		errc := make(chan error)
+		for j := 0; j < testcases; j++ {
+			j := j
+			go func() {
+				err := uploadAndDownload(ctx, sim, nodeCnt, chunkCnt, j)
+				select {
+				case errc <- err:
+				case <-ctx.Done():
+				}
+			}()
+		}
+		i := 0
+		for err := range errc {
+			if err != nil {
+				return err
+			}
+			i++
+			if i >= testcases {
+				return nil
+			}
+		}
+		return nil
+	})
+
+	if result.Error != nil {
+		return fmt.Errorf("simulation error: %v", result.Error)
+	}
+	log.Error("simulation", "duration", time.Since(start))
```

Comment:
`info` or remove altogether (`go test` should give you the time it took)

Reply:
no this is meant to measure the simulation only discounting the time it took the snapshot to be loaded. I noticed that snapshot loading is flaky